### PR TITLE
Privatelink connection

### PR DESCRIPTION
**Pull request by alex-hunt-materialize**
Allows the creation of AWS PrivateLink connections, and the associated K8S VpcEndpoint objects.

Does not include modifications to postgres and kafka sources to actually use those PrivateLink connections.

### Motivation

  * This PR adds a known-desirable feature.

Part of https://github.com/MaterializeInc/cloud/issues/4285

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by alex-hunt-materialize** [comment]
Requires https://github.com/MaterializeInc/cloud/pull/4354 , for the permissions to manage these resources, otherwise it crash loops.

File: src/cloud-resources/src/crd/vpc_endpoint.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    use std::path::PathBuf;
+
+    use kube::core::crd::merge_crds;
+    use kube::CustomResourceExt;
+
+    #[test]
+    fn test_vpc_endpoint_crd_matches() {
+        let crd = merge_crds(vec![super::v1::VpcEndpoint::crd()], "v1").unwrap();
+        let crd_json = serde_json::to_string(&serde_json::json!(&crd)).unwrap();
+        let mut exported_crd_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
```
Comment (benesch) [inline src/cloud-resources/src/crd/vpc_endpoint.rs:59]: This is going to hardcode the path on the build machine, which is not what you want, as the tests are copied into and run in a Docker image! You should instead rely on the guarantee that `cargo test` gives you about the CWD: all test binaries are executed with a CWD of the Cargo manifest directory. Our custom Docker test runner provides the same guarantee!

tl;dr if you just use the relative path `src/crd/gen/vpcendpoints.json` it should all just work out.

File: src/adapter/src/coord/sequencer.rs [hunk 2]
```diff
@@ -588,13 +589,29 @@ impl<S: Append + 'static> Coordinator<S> {
...
+                self.secrets_controller
+                    .ensure(connection_gid, &keyset.to_bytes())
+                    .await?;
+
+                ssh.public_keys = Some(keyset.public_keys());
+            }
+            mz_storage::types::connections::Connection::AwsPrivateLink(ref privatelink) => {
+                self.cloud_resource_controller
+                    .as_ref()
+                    .expect("PrivateLink connections are only allowed in cloud.")
```
Comment (jkosh44) [inline src/adapter/src/coord/sequencer.rs:604]: If someone tries to create one of these connections locally will we panic or is an error returned somewhere earlier? I don't think we should panic and instead return some form of error with this message.

File: src/orchestrator-kubernetes/src/cloud_resource_controller.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        for (key, value) in &self.config.service_labels {
+            labels.insert(key.clone(), value.clone());
+        }
+        let vpc_endpoint = VpcEndpoint {
+            metadata: ObjectMeta {
+                labels: Some(labels),
+                name: Some(name.clone()),
+                namespace: Some(self.kubernetes_namespace.clone()),
+                // TODO owner references?
+                //owner_references: todo!(),
```
Comment (jkosh44) [inline src/orchestrator-kubernetes/src/cloud_resource_controller.rs:50]: Did you mean to leave this in? I'm not really sure what it means.

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 4]
```diff
@@ -727,6 +727,37 @@ impl<T: AstInfo> AstDisplay for AwsConnectionOption<T> {
...
+/// An option in a `CREATE CONNECTION...AWS PRIVATELINK`.
+pub struct AwsPrivateLinkConnectionOption<T: AstInfo> {
+    pub name: AwsPrivateLinkConnectionOptionName,
+    pub value: WithOptionValue<T>,
+}
+
+impl<T: AstInfo> AstDisplay for AwsPrivateLinkConnectionOption<T> {
+    fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
+        f.write_node(&self.name);
+        f.write_node(&self.value);
```
Comment (jkosh44) [inline src/sql-parser/src/ast/defs/ddl.rs:756]: Do we want some form of separator in-between these? Like `<name>: <value>`?

Comment (jkosh44) [reply src/sql-parser/src/ast/defs/ddl.rs:756]: Now that I look every other option has the following format: `<name> = <value>`.

File: src/sql-parser/tests/testdata/ddl [hunk 5]
```diff
@@ -374,6 +374,13 @@ CREATE MATERIALIZED VIEW v IN CLUSTER [1] AS SELECT 1
 =>
 CreateMaterializedView(CreateMaterializedViewStatement { if_exists: Error, name: UnresolvedObjectName([Ident("v")]), columns: [], in_cluster: Some(Resolved("1")), query: Query { ctes: [], body: Select(Select { distinct: None, projection: [Expr { expr: Value(Number("1")), alias: None }], from: [], selection: None, group_by: [], having: None, options: [] }), order_by: [], limit: None, offset: None } })
 
+parse-statement
+CREATE CONNECTION privatelinkconn TO AWS PRIVATELINK (SERVICE NAME 'com.amazonaws.vpce.us-east-1.vpce-svc-0e123abc123198abc', AVAILABILITY ZONES ('use1-az1', 'use1-az4'))
```
Comment (jkosh44) [inline src/sql-parser/tests/testdata/ddl:378]: Can you add one more test that uses the `CREATE CONNECTION privatelinkconn FOR ...` syntax?

**Review by jkosh44 — APPROVED** [review]
Mostly LGTM

- I had a couple of small nits.
- I didn't really see any tests where these connections get used, only where we create them and assert they exist. Is it possible to write those tests?
- I believe that this PR adds the first "cloud only" object to the catalog. That's maybe not a huge deal but something we want to probably think about. This creates a divergence between what a catalog could look like locally and what it could look like in cloud. There's a couple of panics that I think would be better as user errors. In theory you could have maybe taken a copy of a cloud stash and started it locally to do some investigating/debugging, however after this change that's no longer the case. Since there's a time constraint for this PR we can probably go ahead and merge this without thinking too hard about this, but I just wanted to point it out.

File: src/adapter/src/coord.rs [hunk 6]
```diff
@@ -665,12 +673,42 @@ impl<S: Append + 'static> Coordinator<S> {
...
                 CatalogItem::Log(_)
                 | CatalogItem::Type(_)
                 | CatalogItem::Func(_)
-                | CatalogItem::Secret(_)
-                | CatalogItem::Connection(_) => {}
+                | CatalogItem::Secret(_) => {}
+            }
+        }
+
+        if let Some(cloud_resource_controller) = &self.cloud_resource_controller {
```
[the remark this answers is not in the document]
Comment (alex-hunt-materialize) [reply src/adapter/src/coord.rs:697]: The cloud resource controller should always exist in the cloud, but isn't available when running outside of kubernetes. In the context of the process orchestrator, we shouldn't be making privatelink connections.

File: src/adapter/src/coord/ddl.rs [hunk 7]
```diff
@@ -398,6 +407,21 @@ impl<S: Append + 'static> Coordinator<S> {
...
+        for vpc_endpoint in vpc_endpoints {
+            if let Err(e) = self
+                .cloud_resource_controller
+                .as_ref()
+                .expect("PrivateLink connections are only allowed in cloud.")
+                .delete_vpc_endpoint(vpc_endpoint)
+                .await
+            {
+                warn!("Dropping VPC Endpoints has encountered an error: {}", e);
+                // TODO reschedule this?
```
[the remark this answers is not in the document]
Comment (alex-hunt-materialize) [reply src/adapter/src/coord/ddl.rs:420]: Good call, I've created https://github.com/MaterializeInc/cloud/issues/4407

File: src/cloud-resources/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+//! Abstractions for management of PrivateLink VpcEndpoint objects
+
+use std::collections::HashSet;
+use std::fmt::Debug;
+
+use async_trait::async_trait;
+
+use mz_repr::GlobalId;
+//use mz_storage::types::connections::AwsPrivateLinkConnection;
```
[the remark this answers is not in the document]
Comment (alex-hunt-materialize) [reply src/cloud-resources/src/lib.rs:18]: I did not. I'll remove it.

File: src/cloud-resources/src/crd/gen/vpcendpoints.json [hunk 9]
```diff
@@ -0,0 +1 @@
+{"apiVersion":"apiextensions.k8s.io/v1","kind":"CustomResourceDefinition","metadata":{"name":"vpcendpoints.materialize.cloud"},"spec":{"group":"materialize.cloud","names":{"categories":[],"kind":"VpcEndpoint","plural":"vpcendpoints","shortNames":["vpce"],"singular":"vpcendpoint"},"scope":"Namespaced","versions":[{"additionalPrinterColumns":[{"description":"Name of the VPC Endpoint Service to connect to.","jsonPath":".spec.awsServiceName","name":"AwsServiceName","priority":1,"type":"string"},{"description":"Availability Zone IDs","jsonPath":".spec.availabilityZoneIds","name":"AvailabilityZoneIDs","priority":1,"type":"string"}],"name":"v1","schema":{"openAPIV3Schema":{"description":"Auto-generated derived type for VpcEndpointSpec via `CustomResource`","properties":{"spec":{"properties":{"availabilityZoneIds":{"items":{"type":"string"},"type":"array"},"awsServiceName":{"type":"string"}},"required":["availabilityZoneIds","awsServiceName"],"type":"object"},"status":{"nullable":true,"type":"object"}},"required":["spec"],"title":"VpcEndpoint","type":"object"}},"served":true,"storage":true,"subresources":{"status":{}}}]}}
```
[the remark this answers is not in the document]
Comment (alex-hunt-materialize) [reply src/cloud-resources/src/crd/gen/vpcendpoints.json:1]: This is the generated CRD definition which is derived from the spec in `src/cloud-resources/src/crd/vpc_endpoints.rs`. It is used in `cloudtest` to register the CRD, since the environment-controller (which would normally register the CRD) doesn't exist in `cloudtest`.

We enforce that it matches the rust code with a cargo test.

File: src/orchestrator-kubernetes/src/cloud_resource_controller.rs (hunk 3 above)
Comment (alex-hunt-materialize) [reply src/orchestrator-kubernetes/src/cloud_resource_controller.rs:50]: I've created https://github.com/MaterializeInc/cloud/issues/4408 to track that.

Owner references are K8S's way of tying objects with their parents. I think we should probably add these to everything created by environmentd, but it's not critical, since they _should_ be cleaned up when the namespace is deleted.

File: src/sql-parser/tests/testdata/ddl (hunk 5 above)
Comment (alex-hunt-materialize) [reply src/sql-parser/tests/testdata/ddl:378]: Will do.

File: src/sql-parser/src/ast/defs/ddl.rs (hunk 4 above)
Comment (alex-hunt-materialize) [reply src/sql-parser/src/ast/defs/ddl.rs:756]: If I add the equals, then it fails to reparse in the cargo tests. I can add the equals sign if I make it `Option`al, but it seems weird to make something optional when it's required.

```
reparse failed: Expected option value, found equals sign
```

File: src/adapter/src/coord/sequencer.rs (hunk 2 above)
Comment (alex-hunt-materialize) [reply src/adapter/src/coord/sequencer.rs:604]: I'll switch to `AdapterError::Unsupported` here, but at startup we aren't allowed to throw any errors at that point. It's probably ok to `unwrap` there, though, since they shouldn't have any PrivateLink connections at startup since we prevent that here.

**Comment by alex-hunt-materialize** [comment]
>I didn't really see any tests where these connections get used, only where we create them and assert they exist. Is it possible to write those tests?

There's no code to actually use them yet. That's my next task after this is in, to add support to the postgres and kafka connections to go over the privatelink tunnel.

I do have a cloudtest test that it creates the K8S VpcEndpoint object, though.

>I believe that this PR adds the first "cloud only" object to the catalog. That's maybe not a huge deal but something we want to probably think about. This creates a divergence between what a catalog could look like locally and what it could look like in cloud. There's a couple of panics that I think would be better as user errors. In theory you could have maybe taken a copy of a cloud stash and started it locally to do some investigating/debugging, however after this change that's no longer the case. Since there's a time constraint for this PR we can probably go ahead and merge this without thinking too hard about this, but I just wanted to point it out.

We could probably make a dummy implementer of `CloudResourceController`, but when we start actually using the connections, we'll likely need an actual tunnel and DNS provider. That seems very out of scope for this.

File: src/sql-parser/src/ast/defs/ddl.rs (hunk 4 above)
Comment (jkosh44) [reply src/sql-parser/src/ast/defs/ddl.rs:756]: @alex-hunt-materialize I was just thinking about this, shouldn't there at least be a space separating the name and value?
